### PR TITLE
Show repo init settings

### PR DESCRIPTION
**Pull request by timbru**
Added to the repo stats endpoint (/stats/repo). Example output:

```
{
  "publishers": {
    "testbed": {
      "objects": 3,
      "size": 3762,
      "manifests": [
        {
          "uri": "rsync://localhost/repo/testbed/0/217876AA966B965A6EDFDED43C469B90ED11AC6D.mft",
          "this_update": "2022-09-13T15:10:12Z",
          "next_update": "2022-09-14T18:18:12Z"
        }
      ]
    },
    "ca": {
      "objects": 5,
      "size": 8247,
      "manifests": [
        {
          "uri": "rsync://localhost/repo/ca/0/CC2487CF3A9C774BFAE2DCE4DD8368441C75C720.mft",
          "this_update": "2022-09-13T15:10:12Z",
          "next_update": "2022-09-14T16:01:12Z"
        }
      ]
    },
    "ta": {
      "objects": 3,
      "size": 3663,
      "manifests": [
        {
          "uri": "rsync://localhost/repo/ta/0/0A6EA673F04AAA345D605A38E7710A2D8413B56C.mft",
          "this_update": "2022-09-13T15:10:12Z",
          "next_update": "2022-09-14T17:02:12Z"
        }
      ]
    }
  },
  "session": "0102f89f-3639-40cb-a967-68789c7da891",
  "serial": 25,
  "last_update": "2022-09-13T15:15:13.427709Z",
  "sia_base": "rsync://localhost/repo/",
  "rrdp_notification_uri": "https://localhost:3000/rrdp/notification.xml"
}
```

### CODE REVIEW

**Comment by ximon18** [comment]
Is this about `sia_base` and `rrdp_notification_uri`?

What's the reason for this PR?

File: src/pubd/repository.rs [hunk 1]
```diff
@@ -364,18 +343,45 @@ impl RepositoryContent {
...
             self.rrdp.publish(delta, jail, config)?;
-            self.stats.remove_publisher(name, self.rrdp.notification());
-
-            self.write_repository(config)
-        } else {
-            // nothing to remove
-            Ok(())
+            self.write_repository(config)?;
         }
+        // remove publisher is present
```
Comment (ximon18) [inline src/pubd/repository.rs:354]: s/is/if/ present?

File: src/pubd/repository.rs [hunk 2]
```diff
@@ -364,18 +343,45 @@ impl RepositoryContent {
...
+        self.publishers.remove(name);
```
Comment (ximon18) [inline src/pubd/repository.rs:355]: This looks like a logic fix unrelated to the PR? Is that correct?

File: src/pubd/repository.rs [hunk 3]
```diff
@@ -1285,7 +1302,7 @@ impl RepoStats {
     pub fn stale_publishers(&self, seconds: i64) -> Vec<PublisherHandle> {
         let mut res = vec![];
         for (publisher, stats) in self.publishers.iter() {
-            if let Some(update_time) = stats.last_update {
+            if let Some(update_time) = stats.last_update() {
```
Comment (ximon18) [inline src/pubd/repository.rs:1305]: This looks like a logic fix unrelated to the PR? Is that correct?

File: src/pubd/repository.rs [hunk 4]
```diff
@@ -1311,11 +1328,14 @@ impl RepoStats {
 
 impl fmt::Display for RepoStats {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "RSYNC BASE:        {}", self.sia_base)?;
+        writeln!(f, "RRDP NOTIFICATION: {}", self.rrdp_notification_uri)?;
```
Comment (ximon18) [inline src/pubd/repository.rs:1332]: Why not "URI" ?

File: src/pubd/repository.rs [hunk 5]
```diff
@@ -1311,11 +1328,14 @@ impl RepoStats {
 
 impl fmt::Display for RepoStats {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "RSYNC BASE:        {}", self.sia_base)?;
```
Comment (ximon18) [inline src/pubd/repository.rs:1331]: Why "RSYNC" and not "SIA" at in the stats endpoint?

File: src/pubd/repository.rs [hunk 6]
```diff
@@ -1363,17 +1377,81 @@ impl PublisherStats {
...
     }
 
+    /// Returns the most recent "this_update" time
+    /// from all manifest(s) published by this publisher,
+    /// if any.. i.e. there may be 0, 1 or many manifests
     pub fn last_update(&self) -> Option<Time> {
-        self.last_update
+        let mut last_update = None;
+        for mft in self.manifests() {
+            if let Some(seen_update) = last_update {
```
Comment (ximon18) [inline src/pubd/repository.rs:1386]: Introducing a new `seen_update` is confusing, why not just shape the existing `last_update`? Ah I see, because of the assignment... still I found it confusing.

File: src/pubd/repository.rs [hunk 7]
```diff
@@ -1363,17 +1377,81 @@ impl PublisherStats {
...
+    pub fn next_update(&self) -> Time {
+        self.next_update
+    }
+}
+
+impl TryFrom<&Manifest> for PublisherManifestStats {
+    type Error = ();
+
+    fn try_from(mft: &Manifest) -> Result<Self, Self::Error> {
+        let uri = mft.cert().signed_object().cloned().ok_or(())?;
```
Comment (ximon18) [inline src/pubd/repository.rs:1450]: What will `uri` be here if this fails?

File: src/upgrades/pre_0_9_0/pubd_objects_migration.rs [hunk 8]
```diff
@@ -331,9 +326,6 @@ struct OldRepository {
 
     rrdp: OldRrdpServer,
     rsync: RsyncdStore,
-
-    #[serde(default = "RepoStats::default")]
-    stats: RepoStats,
```
Comment (ximon18) [inline src/upgrades/pre_0_9_0/pubd_objects_migration.rs:336]: Why is the serialization of `OldRepository` being changed?

**Review by ximon18 — COMMENTED** [review]
Quick review done on my phone, but ideally I'd review it on a bigger screen, it's a bit hard to follow on a phone. I haven't tested it, would also be good to do that. I don't see any tests updated in this PR to follow the changes... is that correct? Also, am I right in thinking a manual update PR doesn't exist yet for these changes to the API (non-breaking I think) and CLI (breaking, if we even give any guarantees about the text format output)?

File: src/pubd/repository.rs (hunk 2 above)
Comment (timbru) [reply src/pubd/repository.rs:355]: Yeah, that is correct. It was leaving the publisher in the map for the RepositoryContent, with no actual published objects. It doesn't really hurt to do so, but I noticed when touching the code, so figured I would just also remove the map entry.

File: src/pubd/repository.rs (hunk 1 above)
Comment (timbru) [reply src/pubd/repository.rs]: fixed

**Comment by timbru** [comment]
> Is this about `sia_base` and `rrdp_notification_uri`?
> 
> What's the reason for this PR?

It was requested by people who use Krill as a Publication Server. It is, was, not possible to query which values were used for `sia_base` and `rrdp_notification_uri` to initialise the server. With this PR we extend the existing `RepoStats` structure - which is currently only being used by the CLI (not the UI) to include this.

I can see how some of the changes to `RepoStats` can be confusing. We were keeping the stats in memory and updating them on access by a publisher. But this does not include any info on the current server settings. So, we would need to get that information dynamically anyway, and could not have sensible defaults for existing values.

In the end I decided to just remove the object from the struct - i.e. it will be ignored when deserializing existing. It is now always calculated dynamically. This adds only minimal overhead as this is queried infrequently.

 But, as a result I made some changes to how the 'last_update' is determined. Instead of setting this on access by a publisher, we now derive this to be the latest 'this_update' time from all _syntactically correct_ manifests found for the publisher.

File: src/upgrades/pre_0_9_0/pubd_objects_migration.rs (hunk 8 above)
Comment (timbru) [reply src/upgrades/pre_0_9_0/pubd_objects_migration.rs:336]: It was just removed, so it will be ignored, because it no longer needs to be included in the new structure for the migration.

File: src/pubd/repository.rs (hunk 3 above)
Comment (timbru) [reply src/pubd/repository.rs:1305]: Because of the inclusion of server information I needed to change to a dynamically calculated structure. This change is a result of that.

File: src/pubd/repository.rs (hunk 6 above)
Comment (timbru) [reply src/pubd/repository.rs]: renamed things more explicitly - hopefully improving readability

File: src/pubd/repository.rs (hunk 7 above)
Comment (timbru) [reply src/pubd/repository.rs:1454]: An `Err(())`. In this case treat this manifest as syntactically incorrect (which it is - I added a comment as well now) - and we do not include it in our stats about this publisher.

Note that other API end-points let operators review the complete set of objects for publishers. This is just a quick overview. Also, incorrect objects should not be found in practice - as Krill and other conformant CAs will include this.

File: src/pubd/repository.rs (hunk 4 above)
Comment (timbru) [reply src/pubd/repository.rs]: I wanted it to be similar to the CLI, but thanks for raising this, because it actually wasn't..

Changed it now to report the rsync and rrdp base URIs as they are also used when initialising the server and renamed the fields appropriately.

File: src/pubd/repository.rs (hunk 5 above)
Comment (timbru) [reply src/pubd/repository.rs]: see other comment

**Comment by timbru** [comment]
re testing: I don't think we need to test the text representation, but the last_updated per publisher stat is actually used and tested wrt child suspension/unsuspension.
